### PR TITLE
Skip @client queries related to apollo-link-state

### LINKED ISSUE #99: apollo-link-state
Now that apollo-link-state allows for local resolvers, this plugin throws errors on those local resolvers.

Can we get an update to pass the root resolver or something into the plugin to let eslint know about the extra information of the schema?

### PR DESCRIPTION
**Pull request by stipsan**
This will not fix #99, but it will at least stop queries that use `@client` from making the linter fail (causing grief for those who want to use both `apollo-link-state` and `eslint-plugin-graphql`).

So queries like this will no longer fail:
```graphql
  query {
    networkStatus @client {
      isConnected
    }
  }
```

However the downside is that if any field in the query uses the `@client` directive the linter will be ok with it, even if there's non-client fields that **should fail**.
Like here:
```graphql
  query {
    networkStatus @client {
      isConnected
    }
    # The typo here will not be detected
    articess {
      id
      title
    }
  }
```

I tried re-implementing the logic used in `apollo-link-state` itself to strip out fields from the query before sending it over the network: https://github.com/apollographql/apollo-link-state/blob/b81edfdd875c1e3f106ac3ecd1a3e5b0c2e7f8e5/packages/apollo-link-state/src/utils.ts#L14-L31
But I hit a `Maximum call stack size exceeded` error. I will look into this, but it could be worth looking into it in another PR after this one. Why?
Any query with `@client` will fail today, so it blocks anyone from combining this linter plugin with `apollo-link-state`, so this PR will not cause a regression for existing users.

TODO:

- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests pass
- [ ] Update CHANGELOG.md with your change
- [ ] If this was a change that affects the external API, update the README

**Pull Request Labels**

- [x] has-reproduction
- [x] feature
- [ ] blocking
- [ ] good first review

### CODE REVIEW

**Comment by jnwng** [comment]
@stipsan thanks for the PR and taking a look at this — i agree that its important that fit well with the `apollo-link-state` paradigm of using client directives. 

unfortunately i think its a little heavy-handed to fail linting for the entire graphql document if even one part of it has a client directive. while its technically "backwards-compatible" i dont think its a great experience because we're silently ignoring the queries. i would be much more comfortable if stripped fields from the query before linting, lets figure out how to get that to work. 

does that sound reasonable?

**Comment by stipsan** [comment]
I agree the best way forward is to make it strip the client parts of the query.
I'm sure the `Maximum call stack` is solvable, it's just a matter of going deeper in the rabbit hole. 

I'll look into it and hopefully it's easier to get it working than I expected 😄

**Comment by tizmagik** [comment]
Hey @stipsan any update on this? Do you have a branch somewhere where the Macimum call stack error is happening? Would be interested in taking a look to see if I could help out 😁

**Comment by stipsan** [comment]
Hey, I’ll make a repro for you later today :) 

I have been so swamped with other things, didn’t mean to drop the ball on this! 😅

**Comment by stipsan** [comment]
@tizmagik I have updated the PR so it no longer has merge conflicts, and pushed changes so that the `Maximum call stack size exceeded` is reproduced.

You should see something like this if you run `npm test` now:
image

File: test/makeProcessors.js [hunk 1]
```diff
@@ -124,4 +124,11 @@ describe('processors', () => {
       });
     });
   });
+  
+  describe('apollo-link-state', () => {
+    it('skips queries that contain the @client directive', () => {
```
Comment (loicplaire) [inline test/makeProcessors.js]: This test will currently always pass. The `KnownDirectives` validator needs to be added to[ the baseConfig.rules object](https://github.com/stipsan/eslint-plugin-graphql/blob/2bd3721a2b68c0444d8efa219e070ec1ef526399/test/makeProcessors.js#L14):

```js
baseConfig: {
  rules: {
    'graphql/required-fields': [
      'error',
      {
        schemaJson,
        env: 'literal',
        requiredFields: ['id'],
        validators: ['KnownDirectives'],
      },
    ],
  },
},
```

**Comment by loicplaire** [comment]
@stipsan have look at this: https://github.com/apollographql/apollo-client/issues/1717#issuecomment-380984397. 

If I add `{ noLocation: true }` here: https://github.com/stipsan/eslint-plugin-graphql/blob/2bd3721a2b68c0444d8efa219e070ec1ef526399/src/index.js#L394, then the test passes. But adding this option make other tests fails...

We could potentially do `ast = parse(text, { noLocation: text.includes('@client') });` but I don't think it make sense to use [that `noLocation` option](https://github.com/graphql/graphql-js/blob/4dc81fb224f7e08f6781a8b1a27a8cafbffb6358/src/language/parser.js#L79) since I am guessing that the linter needs to know the location of the error.

I seems to me that it's an issue with `apollo-utilities` so I wonder if you could potentially use another lib that let's you remove a directive from an AST.

I don't have a solution for this but I hope this sheds some lights on the issue you are having.

**Comment by stipsan** [comment]
I've cloned the `removeDirectivesFromDocument` utility, removing the cloneDeep part and it looks like it is passing now @loicplaire. I failed to find another package on npm that was able to remove parts of the graphql AST.

@jnwng what do you say?

**Comment by jnwng** [comment]
i took a peek at this and the cloneDeep call chokes on the “loc” property
of the document. calling cloneDeep without the “loc” does what we want (i
think) but requires us to either re-parse the document (this is probably
okay) passing the noLoc flag or to remove the loc from our already parsed
document.

in either case, as long as the clone is doing what we want it should be
okay to move forward!

**Comment by stipsan** [comment]
I tried something like:
```
if(hasDirectives(['client'], ast)) {
  const {loc, ...astWithoutLoc} = ast
  ast = removeDirectivesFromDocument( [{ name: 'client' }], astWithoutLoc);
}
```

But it still hit the overflow, I can only assume because of the other `.loc` properties that are all over the tree.

@jnwng what do you think about making a PR on the apollo-utilities project to add a flag to skip the deep clone on `removeDirectivesFromDocument ` so we avoid maintaining all this duplicated code here? I recon we don't really need the clone here since the ast is not used elsewhere so the mutations should be safe in this context 🤔

**Comment by MrTibbles** [comment]
Any movement on this? Feels like plenty of time for thought to be furthered.

**Comment by MrTibbles** [comment]
Resolved, see [comment](https://github.com/apollographql/eslint-plugin-graphql/issues/99#issuecomment-601701257)

**Pull request closed without merging**